application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,120 @@
+package org.togetherjava.tjbot.commands.moderation;
+
+import net.dv8tion.jda.api.Permission;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import net.dv8tion.jda.api.interactions.Interaction;
+import net.dv8tion.jda.api.interactions.commands.OptionMapping;
+import net.dv8tion.jda.api.interactions.commands.OptionType;
+import net.dv8tion.jda.api.requests.RestAction;
+import net.dv8tion.jda.api.utils.Result;
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.SlashCommandAdapter;
+import org.togetherjava.tjbot.commands.SlashCommandVisibility;
+import org.togetherjava.tjbot.config.Config;
+
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+
+/**
+ * This command can warn users. The command will also try to DM the user to inform them about the
+ * action and the reason.
+ * <p>
+ * The command fails if the user triggering it is lacking permissions to either warn other users or
+ * to warn the specific given user (for example a moderator attempting to ban an admin).
+ */
+public final class WarnCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String REASON_OPTION = "reason";
+    private static final String ACTION_VERB = "warn";
+    private final ModerationActionsStore actionsStore;
+    private final Predicate<String> hasRequiredRole;
+
+    /**
+     * Creates a new Instance.
+     *
+     * @param actionsStore used to store actions issued by this command
+     */
+    public WarnCommand(@NotNull ModerationActionsStore actionsStore) {
+        super("warn", "warns the user", SlashCommandVisibility.GUILD);
+
+        getData().addOption(OptionType.USER, USER_OPTION, "The user to warn", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "The reason for the warning", true);
+
+        hasRequiredRole = Pattern.compile(Config.getInstance().getHeavyModerationRolePattern())
+            .asMatchPredicate();
+        this.actionsStore = Objects.requireNonNull(actionsStore);
+    }
+
+    private static @NotNull RestAction<Boolean> dmUser(long userId, @NotNull String reason,
+            @NotNull SlashCommandEvent event) {
+        return event.getJDA()
+            .openPrivateChannelById(userId)
+            .flatMap(channel -> channel.sendMessage(
+                    """
+                            Hey there, We are sorry to inform you that you have been warned for the following reason: %s
+                            The repercussion of this warning depends on the severity of the warning or one the number of warnings you have received.\040
+                            The warn system can be seen in the <#652440333107331082> channel.\040
+                            """
```

Comment:
If I would be a user receiving this message, I would be super confused. Also, it has some typos and that `\040` is also very weird. The channel link probably also does not work outside of the guild (dont forget that this is a DM).

Replace with:
```java
Hey there, sorry to tell you but unfortunately you have been warned in the server %s.
If you think this was a mistake, please contact a moderator or admin of the server.
The reason for the warning is: %s
```